daft/io/sql/sql_data_sink.py

```diff
@@ -0,0 +1,285 @@
+
+        if isinstance(self._connection, str):
+            return create_engine(self._connection).connect()
+        else:
+            return self._connection()
+
+    def _create_table(self) -> None:
+        """Create a new table in the database."""
+        from sqlalchemy import inspect
+
+        try:
+            conn = self._get_connection()
+            try:
+                # Check if table already exists
+                inspector = inspect(conn)
+                if self._table_name in inspector.get_table_names():
+                    if self._mode == "create":
+                        raise ValueError(
+                            f"Table '{self._table_name}' already exists. Use mode='append' or mode='replace'"
+                        )
+            finally:
+                conn.close()
+        except Exception:
+            # If we can't check, we'll let the write fail if needed
+            pass
+
+    def write(self, micropartitions: Iterator[MicroPartition]) -> Iterator[WriteResult[dict[str, Any]]]:
+        """Write micropartitions to the SQL database using PyArrow tables.
+
+        Args:
+            micropartitions: Iterator of micropartitions to write
+
+        Yields:
+            WriteResult: Result containing rows/bytes written metadata
+        """
+        from sqlalchemy import MetaData
+
+        conn = self._get_connection()
+        try:
+            metadata = MetaData()
+            table_exists = False
+            table = None
+
+            for i, micropartition in enumerate(micropartitions):
+                # Convert MicroPartition to PyArrow table
+                arrow_table = micropartition.to_arrow()
+                rows_written = arrow_table.num_rows
+                bytes_written = arrow_table.nbytes
+
+                # On first write, create the table schema if needed
+                if i == 0 and self._mode in ("create", "replace"):
+                    table = self._create_table_from_arrow(conn, arrow_table, metadata)
+                    table_exists = True
+                elif i == 0 and self._mode == "append":
+                    # Load existing table metadata
+                    metadata.reflect(bind=conn)
+                    if self._table_name not in metadata.tables:
+                        raise ValueError(f"Table '{self._table_name}' does not exist. Use mode='create'")
+                    table = metadata.tables[self._table_name]
+                    table_exists = True
+
+                # Insert rows into table using executemany with pylist
+                if table_exists and table is not None:
+                    self._insert_arrow_table(conn, table, arrow_table)
+
+                yield WriteResult(
+                    result={"table_name": self._table_name},
+                    bytes_written=bytes_written,
+                    rows_written=rows_written,
+                )
+
+            # Commit the transaction
+            conn.commit()
```

Comment:
Btw @malcolmgreaves let's also consider starting the transaction in the `start()` then committing in the `finalize()`